irohad/ametsuchi/impl/flat_file/flat_file.cpp

```diff
@@ -16,132 +16,84 @@
+   * DIGIT_CAPACITY-character width regardless of the value of `id`.
+   * If the length of the string representation of `id` is less than
+   * DIGIT_CAPACITY, then the returned value is filled with leading zeros.
+   *
+   * For example, if str_rep(`id`) is "123", then the returned value is
+   * "0000000000000123".
+   *
+   * @param id - for conversion
+   * @return string repr of identifier
+   */
+  std::string id_to_name(Identifier id) {
+    std::ostringstream os;
+    os << std::setw(DIGIT_CAPACITY) << std::setfill('0') << id;
+    return os.str();
   }
-}
 
-/**
- * Checking consistency of storage for provided folder
- * If some block in the middle is missing all blocks following it are deleted
- * @param dump_dir - folder of storage
- * @return - last available identifier
- */
-nonstd::optional<Identifier> check_consistency(const std::string &dump_dir) {
-  auto log = logger::log("FLAT_FILE");
-
-  Identifier tmp_id = 0u;
-  if (dump_dir.empty()) {
-    log->error("check_consistency({}), not directory", dump_dir);
-    return nonstd::nullopt;
-  }
-  // Directory iterator:
-  struct dirent **namelist;
-  auto status = scandir(dump_dir.c_str(), &namelist, nullptr, alphasort);
-  if (status < 0) {
-    log->error("check_consistency({}), scandir error: {}", dump_dir, status);
-    return nonstd::nullopt;
-  }
-  if (status < 3) {
-    log->info("check_consistency({}), directory is empty", dump_dir);
-    auto n = static_cast<uint32_t>(status);
-    for (auto j = 0u; j < n; ++j) {
-      free(namelist[j]);
+  /**
+   * Checking consistency of storage for provided folder
+   * If some block in the middle is missing all blocks following it are deleted
+   * @param dump_dir - folder of storage
+   * @return - last available identifier
+   */
+  nonstd::optional<Identifier> check_consistency(const std::string &dump_dir) {
+    auto log = logger::log("FLAT_FILE");
+
+    if (dump_dir.empty()) {
+      log->error("check_consistency({}), not directory", dump_dir);
+      return nonstd::nullopt;
     }
-    free(namelist);
-    return 0;
-  }
-
-  auto n = static_cast<uint32_t>(status);
-  tmp_id = 2;
-
-  while (tmp_id < n and id_to_name(tmp_id - 1) == namelist[tmp_id]->d_name) {
-    ++tmp_id;
-  }
-  for (auto j = tmp_id; j < n; ++j) {
-    remove(dump_dir, namelist[j]->d_name);
-  }
 
-  for (auto j = 0u; j < n; ++j) {
-    free(namelist[j]);
+    auto const files = [&dump_dir] {
+      std::vector<boost::filesystem::path> ps;
+      std::copy(boost::filesystem::directory_iterator{dump_dir},
+                boost::filesystem::directory_iterator{},
+                std::back_inserter(ps));
+      std::sort(ps.begin(),
+                ps.end(),
+                [](const boost::filesystem::path &lhs,
+                   const boost::filesystem::path &rhs) {
+                  return lhs.compare(rhs) < 0;
+                });
+      return ps;
+    }();
+
+    auto const missing =
```

Comment:
Can be simplified with boost:
```C++
auto const missing = boost::range::find_if(
        files | boost::adaptors::indexed(1), [](const auto &it) {
          return id_to_name(it.index()) != it.value().filename();
        });
```

Reply:
even better to use [binary_search](http://www.boost.org/doc/libs/1_62_0/libs/range/doc/html/range/reference/algorithms/non_mutating/binary_search.html), as long as you sort it before

Reply:
Unfortunately, std or boost binary_search does not returns the position of the found object, which I use to calculate the valid index at the return statement. We can introduce a local variable and capture the reference to it. But I'm not sure if doing so can buy us much here, since here file IO is much expensive than searching.

Reply:
Regarding to use boost::adaptors::indexed, using the adaptors we lose the composability because the adaptor change the type of the returned value.

For example, I could not find the way to call for_each with the adaptor applied. If we can find the way I can use the adaptor.

I need to find the way to replace the following operations after the search with adaptor:

- calling for_each to remove files after the missing IDs.
- compute the last valid ID by "return missing - files.cbegin()"

Reply:
Calling `.get()` on adaptor returns original iterator, so it can be added either after `find_if()`, or when `missing` is accessed.

Reply:
Thanks! It worked!